wallet/manager.go

```diff
@@ -237,32 +270,244 @@ func (m *Manager) UnconfirmedEvents() ([]Event, error) {
+			}
+		}
 	}
 
-	// sleep for the duration and then unreserve the ids
-	time.AfterFunc(duration, func() {
-		m.mu.Lock()
-		defer m.mu.Unlock()
+	tip := m.chain.Tip()
+	if amount.IsZero() {
+		return nil, tip, types.ZeroCurrency, nil
+	}
 
-		for _, id := range ids {
-			delete(m.used, id)
+	var inputSum types.Currency
+	var selected []types.SiacoinElement
+	var utxoIDs []types.Hash256
+	const utxoBatchSize = 100
+top:
+	for i := 0; ; i += utxoBatchSize {
+		// extra large wallets may need to paginate through utxos
+		// to find enough to cover the amount
+		utxos, err := m.store.WalletSiacoinOutputs(walletID, tip, i, utxoBatchSize)
+		if err != nil {
+			return nil, types.ChainIndex{}, types.ZeroCurrency, fmt.Errorf("failed to get siacoin elements: %w", err)
+		} else if len(utxos) == 0 {
+			return nil, types.ChainIndex{}, types.ZeroCurrency, ErrInsufficientFunds
 		}
-	})
-	return nil
+
+		for _, sce := range utxos {
+			if inPool[sce.ID] || m.utxosLocked(types.Hash256(sce.ID)) != nil {
+				continue
+			}
+
+			selected = append(selected, sce)
+			utxoIDs = append(utxoIDs, types.Hash256(sce.ID))
+			inputSum = inputSum.Add(sce.SiacoinOutput.Value)
+			if inputSum.Cmp(amount) >= 0 {
+				break top
+			}
+		}
+	}
+
+	if inputSum.Cmp(amount) < 0 {
+		if !useUnconfirmed {
+			return nil, types.ChainIndex{}, types.ZeroCurrency, ErrInsufficientFunds
+		}
+
+		for _, sce := range ephemeral {
+			if inPool[sce.ID] || m.utxosLocked(types.Hash256(sce.ID)) != nil {
+				continue
+			}
+
+			selected = append(selected, sce)
+			inputSum = inputSum.Add(sce.SiacoinOutput.Value)
+			if inputSum.Cmp(amount) >= 0 {
+				break
+			}
+		}
+	}
+
+	if inputSum.Cmp(amount) < 0 {
+		return nil, types.ChainIndex{}, types.ZeroCurrency, ErrInsufficientFunds
+	}
+	m.lockUTXOs(utxoIDs...)
+	return selected, tip, inputSum.Sub(amount), nil
+}
+
+// SelectSiafundElements selects siafund elements from the wallet that sum to
+// at least the given amount. Returns the elements, the element basis, and the
+// change amount.
+func (m *Manager) SelectSiafundElements(walletID ID, amount uint64) ([]types.SiafundElement, types.ChainIndex, uint64, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	tip := m.chain.Tip()
+	if amount == 0 {
+		return nil, tip, 0, nil
+	}
+
+	knownAddresses := make(map[types.Address]bool)
+	relevantAddr := func(addr types.Address) (bool, error) {
+		if exists, ok := knownAddresses[addr]; ok {
+			return exists, nil
+		}
+		_, err := m.store.WalletAddress(walletID, addr)
+		if errors.Is(err, ErrNotFound) {
+			knownAddresses[addr] = false
+			return false, nil
+		} else if err != nil {
+			return false, err
+		}
+		knownAddresses[addr] = true
+		return true, nil
+	}
+
+	ephemeral := make(map[types.SiafundOutputID]types.SiafundElement)
+	inPool := make(map[types.SiafundOutputID]bool)
+	for _, txn := range m.chain.PoolTransactions() {
+		for _, sfi := range txn.SiafundInputs {
+			inPool[sfi.ParentID] = true
+			delete(ephemeral, sfi.ParentID)
```

Comment:
This `for` loop deletes ephemeral elements, but they're getting populated in the next loop, doesn't that mean we might miss deleting an ephemeral element if the same transaction holds both?  Might be an edge case that's impossible, not sure.

Reply:
ephemeral elements can't be spent before they're created, that should never happen.